lib/Sema/TypeCheckType.cpp

```diff
@@ -6079,12 +6079,23 @@ class ExistentialTypeSyntaxChecker : public ASTWalker {
       // If this is a type alias to a constraint type, the type
       // alias name must be prefixed with 'any' to be used as an
       // existential type.
-      if (type->isConstraintType()) {
+      if (type->isConstraintType() && !type->isAny() && !type->isAnyObject()) {
+        bool diagnose = false;
+
+        // Look for protocol members that require 'any'.
         auto layout = type->getExistentialLayout();
         for (auto *protoDecl : layout.getProtocols()) {
-          if (!protoDecl->existentialRequiresAny())
-            continue;
+          if (protoDecl->existentialRequiresAny()) {
+            diagnose = true;
+            break;
+          }
+        }
+
+        // If inverses are present, require 'any' too.
+        if (auto *PCT = type->getAs<ProtocolCompositionType>())
```

Comment:
This either needs getCanonicalType() or walk into nested PCTs; PCT::get() doesn't flatten unless the type is going to be canonical.

At that point you may as well walk the members of the canonical PCT directly instead of consulting the existential layout. Then you won't need the hard-coded isAny/isAnyObject() condition.

Reply:
Thanks. There's also a number of bugs around parsing of things like `borrowing ~Copyable` so I'll file a separate bug to track this fix too.